linenoise.c

```diff
@@ -794,10 +824,54 @@ static int linenoiseEdit(int stdin_fd, int stdout_fd, char *buf, size_t buflen,
     if (write(l.ofd,prompt,l.plen) == -1) return -1;
     while(1) {
         char c;
-        int nread;
+        int nread = 0;
         char seq[3];
 
-        nread = read(l.ifd,&c,1);
+        fd_set rfds;
+        int retval;
+
+        FD_ZERO(&rfds);
+        FD_SET(l.ifd, &rfds);
+        FD_SET(ctrl_pipe_read_fd, &rfds);
+        int nfds = (l.ifd > ctrl_pipe_read_fd) ? l.ifd+1 : ctrl_pipe_read_fd+1; //compute 1+(the greater fd) for select
+
+        struct timeval timeout;
+        timeout.tv_sec = 0;
+        timeout.tv_usec = 1e5; // 0.1 seconds
+
+        retval = select(nfds, &rfds, NULL, NULL, delayedRefresh ? &timeout : NULL);
+        if (retval == -1)
+        {
+            fprintf(stderr, "Problem with select in linenoise: %s\n", strerror(errno));
+            exit(errno);
+        }
+        else if (retval)
+        {
+            if (FD_ISSET(l.ifd, &rfds))
+            {
+                nread = read(l.ifd,&c,1);
+            }
+            else if (FD_ISSET(ctrl_pipe_read_fd, &rfds))
+            {
+                nread = read(ctrl_pipe_read_fd,&c,1);
+            }
+            else
+            {
+                /* don't know how we would get here */
+                continue;
+            }
+        }
+        else
+        {
+            // timeout
+            if (delayedRefresh)
+            {
+                delayedRefresh = 0;
+                refreshLine(&l);
+            }
+            continue;
+        }
+
```

Comment:
Should you have a `delayedRefresh = 0` here as well, otherwise typing at the prompt will not prevent it from happening just put it off until you're done typing really fast...